projects/laji/src/app/shared-modules/observation-map/observation-map/observation-map.component.ts

```diff
@@ -209,210 +237,272 @@ export class ObservationMapComponent implements OnChanges, OnDestroy {
-            properties[name] = ObservationMapComponent.getValue(row, field);
-          });
-          features.push({
-            type: 'Feature',
-            geometry: {
-              type: 'Point',
-              coordinates
-            },
-            properties
-          });
-        });
+  private prepareQuery(): WarehouseQueryInterface {
+    const query = { ...this.query };
+
+    this.addVisualizationParams(query);
+
+    if (this.useFinnishMap) {
+      if (!query.coordinates) {
+        query.coordinates = FINNISH_MAP_BOUNDS;
       }
-      return {
-        lastPage: 1,
-        featureCollection: {
-          type: 'FeatureCollection',
-          features
-        },
-        cluster: {
-          spiderfyOnMaxZoom: true,
-          showCoverageOnHover: true,
-          singleMarkerMode: true,
-          maxClusterRadius: 20
-        }
-      };
-    })).pipe(tap(() => {
-      this.showingIndividualPoints = true;
-    }));
+    } else {
+      this.addViewPortCoordinatesParams(query);
+    }
+
+    return query;
   }
 
-  private getCountForQuery$(query: WarehouseQueryInterface) {
-    return this.warehouseService.warehouseQueryCountGet(query).pipe(
-      map(result => result.total)
+  private getDrawData$(query: WarehouseQueryInterface): Observable<LajiMapDataOptions> {
+    return getFeatureCollectionFromQueryCoordinates$(
+      query.coordinates, this.useFinnishMap
+    ).pipe(
+      tap(featureCollection => {
+        this.drawData = {...this.drawData, featureCollection};
+      }),
+      map(() => this.drawData)
     );
   }
 
-  private tryToQueryIndividualPoints$(query: WarehouseQueryInterface, page: number) {
-    return this.getCountForQuery$(query).pipe(
-      switchMap(cnt => {
-        if (!cnt) {
-          return of({
-            lastPage: 1,
-            featureCollection: {
-              type: 'FeatureCollection',
-              features: []
-            }
-          });
-        } else if (cnt < this.showIndividualPointsWhenLessThan) {
-          return this.getItemsAsPoints$(query);
+  private getDataOptions$(query: WarehouseQueryInterface): Observable<LajiMapDataOptions> {
+    this.loading = true;
+    this.cdr.markForCheck();
+
+    return this.getPointsDataOptions$(query).pipe(
+      switchMap(pointsDataOptions => {
+        if (pointsDataOptions.total <= this.pointGeometryPageSize) {
+          return of(pointsDataOptions).pipe(tap(() => this.showingIndividualPoints = true));
         } else {
-          return (this.warehouseService.warehouseQueryAggregateGet(
-            query, this.activeZoomThresholdLevelToAggregateBy(),
-            undefined, this.size, page, true
-          )).pipe(tap(() => {
-            this.showingIndividualPoints = false;
-          }));
+          return this.getBoxDataOptions$(query).pipe(tap(() => this.showingIndividualPoints = false));
         }
-      }));
+      }),
+      // retry on timeout
+      timeout(WarehouseApi.longTimeout * 3),
+      delay(100),
```

Comment:
What's the delay about? If it's necessary could you add some explanation into a comment for it.

Reply:
I actually don't know, I didn't change that line because I didn't know why it was there (git shows it as new line for some reason)